spec.md

```diff
@@ -0,0 +1,81 @@
+- Add a new error `Log` to the `Span`; analagous to adding an info log aside from the severity. **(py: `error`, go: `Error`)**
+- Access the `TraceContext` associated with the `Span`. **(py: `trace_context`, go: `TraceContext()`)**
+
+
+#### TraceContext
+
+Every `TraceContext` must provide access to "trace attributes". A **trace attribute** is a key:value pair associated with a TraceContext **that also propagates to future TraceContext children** per TraceContext.NewChild, and that propagates across process boundaries in-band with application data.
+
+Trace attributes are powerful, especially given an OpenTracing integration that extends across many layers in a distributed stack. For example, arbitrary application data from a mobile app can travel, transparently, all the way into the depths of a storage system). These powerful capabilities come with some powerful costs: use this feature with care.
+
+Also, trace attribute keys have a restricted format: implementations may wish to use them as HTTP header keys (or key suffixes), and of course HTTP headers are case insensitive.
+
+As such, trace attribute keys MUST match the regular expression `(?i:[a-z0-9][-a-z0-9]*)`, and – per the `?i:` – they are case-insensitive.  That is, the trace attribute key must start with a letter or number, and the remaining characters must be letters, numbers, or hyphens. See CanonicalizeTraceAttributeKey(). If a trace attribute key does not meet these criteria, SetTraceAttribute() results in undefined behavior.
+
+In any case, the programmatic `TraceContext` interface is deceptively simple:
+
+- Set a trace attribute, which is a simple string:string pair. Note that newly-set trace attributes are only guaranteed to propagate to *future* children of the `TraceContext` and/or its associated `Span`. **(py: `set_trace_attribute`, go: `SetTraceAttribute`)**
+- Get a trace attribute by key. **(py: `get_trace_attribute`, go: `TraceAttribute`)**
+
+
+#### TraceContextSource, TraceContextEncoder, TraceContextDecoder
+
+Implementations create `TraceContext` instances via a `TraceContextSource` interface; it must have the following capabilities:
+
+- Create a new parent-less ("root") `TraceContext`. **(py: `new_root_trace_context`, go: `NewRootTraceContext`)**
+- Create a new child `TraceContext` given a parent `TraceContext`. **(py: `new_child_trace_context`, go: `NewChildTraceContext`)**
+
+
+Additionally, a `TraceContextSource` must be able to encode and decode `TraceContext`s. In its encoded form, a `TraceContext` is separated into a *pair* of values: one represents the "span identity" – for example, in Zipkin or Dapper, this would be the `trace_id` and `span_id`; and the other encoded value represents the trace attributes. This separation of encoded state enables optimizations in certain binary protocols.
```

Comment:
Should make it clear here that the encoding is not standardized anywhere, i.e. not required to be compatible with other tracing systems.